### ISSUE TITLE
[megatron][engine][context parallel] batch_num_tokens aggregation undercounts valid tokens when CP > 1

### ISSUE THREAD

**Issue by alex27594** [verl-project/verl#5983]
### Summary

In the Megatron engine path, `batch_num_tokens` appears to be aggregated over the pure DP group in `transformer_impl.py`.

When `CP > 1`, each CP rank only holds a 1/CP slice of the sequence, so reducing `batch_num_tokens` over pure DP undercounts valid tokens by a factor of `CP`.

This affects `loss_agg_mode="token-mean"` normalization in the engine path.

### Current behavior

For `CP > 1`, `batch_num_tokens` is effectively smaller than the true valid-token count for the full DP×CP shard.

As a result, token-mean normalization is incorrect.

### Expected behavior

For `CP > 1`, `batch_num_tokens` should be reduced over the DP×CP group, while `dp_size` should remain pure DP.

### Proposed fix

Use `mpu.get_data_parallel_group(with_context_parallel=True)` when reducing `batch_num_tokens` in the Megatron engine path.

### Validation

I have:
- a minimal code fix in `transformer_impl.py`
- a focused test / reproducer for the CP>1 case
- manual smoke coverage for the engine path

### Related context

Related to #4852  
Potentially related to #1332

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The bug can be reproduced with a minimal Megatron engine-path setup where `context_parallel_size > 1`.

#### Config

- `tensor_model_parallel_size = 1`
- `pipeline_model_parallel_size = 1`
- `context_parallel_size = 2`
- `data_parallel_size = 1` (minimal case)
- `loss_agg_mode = "token-mean"`

#### Minimal reproducer

```python
import torch
import torch.distributed as dist
from megatron.core import parallel_state as mpu

def main():
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    local_rank = int(torch.cuda.current_device())
    device = torch.device("cuda", local_rank)

    # Assume Megatron parallel state has already been initialized with:
    # TP=1, PP=1, CP=2
    #
    # Each CP rank only sees a 1/CP slice of the sequence.
    # For a fully valid local slice:
    loss_mask = torch.ones(8, device=device)  # 8 valid tokens on each CP rank
    local_tokens = loss_mask.sum().to(torch.int64)  # = 8 on each rank

    # Current behavior in the buggy engine path:
    wrong = local_tokens.clone()
    dist.all_reduce(
        wrong,
        op=dist.ReduceOp.SUM,
        group=mpu.get_data_parallel_group(),   # pure DP group
    )

    # Expected behavior:
    correct = local_tokens.clone()
    dist.all_reduce(
        correct,
        op=dist.ReduceOp.SUM,
        group=mpu.get_data_parallel_group(with_context_parallel=True),  # DP×CP group
    )

    if rank == 0:
        print(f"local_tokens={int(local_tokens)}")
        print(f"wrong_batch_num_tokens={int(wrong)}")
        print(f"correct_batch_num_tokens={int(correct)}")

        # With CP=2, wrong_batch_num_tokens is smaller by a factor of 2.
        # This undercounts valid tokens and makes token-mean normalization incorrect.

if __name__ == "__main__":
    main()
```

### Expected behavior

For `CP > 1`, the Megatron engine path should aggregate `batch_num_tokens` over the correct DP×CP group, while keeping `dp_size` equal to pure DP. This should make `loss_agg_mode="token-mean"` use the correct valid-token count and avoid CP-dependent undercounting.

**Comment by ianliuy** [comment]
## What's broken?

When using the Megatron engine with context parallelism (CP > 1), `batch_num_tokens` is undercounted by a factor of CP. This causes `loss_agg_mode="token-mean"` to produce an inflated loss value, leading to incorrect gradient scaling and potentially affecting training convergence.

## Who is affected?

Any user running Megatron-backend training with `context_parallel_size > 1` and `loss_agg_mode="token-mean"` (the default for SFT and used in PPO/GRPO). Users with `context_parallel_size = 1` are **not** affected. This likely explains the convergence issues reported in #1332 (GRPO + CP > 1).

## When does it trigger?

Every training step when CP > 1. The bug is deterministic, not intermittent. It can be reproduced with any Megatron engine config where `context_parallel_size > 1`.

## Where is the bug?

`verl/workers/engine/megatron/transformer_impl.py`, in `forward_backward_batch()`:

```python
# Line ~596-602
batch_num_tokens = data["loss_mask"].sum().to(get_device_id())
torch.distributed.all_reduce(
    batch_num_tokens, op=torch.distributed.ReduceOp.SUM,
    group=self.get_data_parallel_group()  # pure DP group, excludes CP ranks
)
tu.assign_non_tensor(data, batch_num_tokens=batch_num_tokens.item())
tu.assign_non_tensor(data, dp_size=self.get_data_parallel_size())  # also pure DP
```

## Why does it happen?

With CP > 1, each CP rank only holds a `1/CP` slice of the full sequence. `loss_mask.sum()` on each CP rank gives a partial token count. The `all_reduce` over the **pure DP group** only sums across DP replicas — it never includes the other CP ranks' partial counts.

This was introduced in commit `b49178f` (#3994), which correctly implemented token-mean normalization for pure DP but predated full CP integration.

## How to fix?

Use `mpu.get_data_parallel_group(with_context_parallel=True)` and `mpu.get_data_parallel_world_size(with_context_parallel=True)` for both `batch_num_tokens` and `dp_size`. Dynamic context parallelism is guarded to preserve existing behavior:

```python
if self.engine_config.dynamic_context_parallel:
    loss_reduce_group = self.get_data_parallel_group()
    dp_size = self.get_data_parallel_size()
else:
    loss_reduce_group = mpu.get_data_parallel_group(with_context_parallel=True)
    dp_size = mpu.get_data_parallel_world_size(with_context_parallel=True)
```

## How do we know it works?

CPU-only regression test validates DP+CP group gives correct total token count with CP=2; CP=1 no-regression test passes.

---

Implemented as described above. PR: #6049
